### ISSUE TITLE
False positive `unresolved-attribute` when creating nested Pydantic models with higher-order functions

### ISSUE THREAD

**Issue by mariuswinger** [astral-sh/ty#4515]
### Summary

Discovered what seems to be a false positive `unresolved-attribute` error when running ty on an example combining pydantic and higher order functions.

## Minimal reproduction

```py
from __future__ import annotations

from pydantic import BaseModel


class Foo(BaseModel):
    start: int
    end: int

    @property
    def midpoint(self) -> float:
        return (self.end + self.start) / 2


class Bar(BaseModel):
    foos: list[Foo]

    def filter_foos(self, limit: float) -> Bar:
        return Bar(foos=list(filter(lambda foo: foo.midpoint <= limit, self.foos)))

    def filter_foos_b(self, limit: float) -> Bar:
        return Bar(foos=list(filter(lambda foo: limit >= foo.midpoint, self.foos)))
```

Running `ty check example.py` produces
```
error[unresolved-attribute]: Attribute `midpoint` is not defined on `Mapping[str, Any]` in union `Foo | Mapping[str, Any]`
  --> src/example.py:22:58
   |
22 |         return Bar(foos=list(filter(lambda foo: limit >= foo.midpoint, self.foos)))
   |                                                          ^^^^^^^^^^^^
```
but in this case we know that `self.foos` is of type `list[Foo]`, and not the broader union type, meaning the `midpoint` property should be guaranteed to exist.
This only happens when putting the higher order call result directly into the pydantic model init. Likely because pydantic are able to coerce `Mapping[str, Any]` into the actual model and therefore accepts the union type in its init. As the example shows, it is also seems to be dependent on the order of comparison in this case, which I have no explanation for why.
Also found the same issues when e.g. using `dropwhile` from itertools instead of `filter`.

## Environment: 
Python 3.12.12
pydantic 2.13.5
No other relevant sections from `pyproject.toml`

### Version

ty 0.0.80 (7fd8e1569 2026-09-09)

**Comment by sharkdp** [comment]
Thank you for reporting this. This is not specifically related to Pydantic. We currently track this here: https://github.com/astral-sh/ty/issues/4016
